File: app/src/main/java/com/dpis/module/applist/AppListFilterState.java
package com.dpis.module.applist;

/** Immutable catalogue filter selection. Type chips are independent; configuration chips are additive. */
public final class AppListFilterState {
    public enum AppType { USER, SYSTEM, ALL }
    public enum ConfigurationFilter { ALL, INJECTED, DISABLED, VIEWPORT, FONT, TYPEFACE, HOOK }
    public enum SortOrder { NAME, UPDATED, INSTALLED }
    private final boolean allAppsSelected, userAppsSelected, systemAppsSelected, injectedOnly, disabledOnly, widthConfiguredOnly, fontConfiguredOnly, typefaceConfiguredOnly, hookConfiguredOnly, reverseOrder;
    private final SortOrder sortOrder;
    public AppListFilterState(AppType type, boolean injected, boolean viewport, boolean font, SortOrder sort, boolean reverse) {
        this(type == AppType.ALL, type != AppType.SYSTEM && type != AppType.ALL, type == AppType.SYSTEM,
                injected, false, viewport, font, false, false, sort, reverse);
    }
    public AppListFilterState(AppType type, boolean injected, boolean disabled, boolean viewport, boolean font,
                              boolean typeface, boolean hook, SortOrder sort, boolean reverse) {
        this(type == AppType.ALL, type != AppType.SYSTEM && type != AppType.ALL, type == AppType.SYSTEM,
                injected, disabled, viewport, font, typeface, hook, sort, reverse);
    }
    public AppListFilterState(boolean user, boolean system, ConfigurationFilter config, SortOrder sort, boolean reverse) {
        this(false, user, system, config == ConfigurationFilter.INJECTED, config == ConfigurationFilter.DISABLED,
                config == ConfigurationFilter.VIEWPORT, config == ConfigurationFilter.FONT,
                config == ConfigurationFilter.TYPEFACE, config == ConfigurationFilter.HOOK, sort, reverse);
    }
    public AppListFilterState(boolean all, boolean user, boolean system, boolean injected, boolean disabled,
                               boolean viewport, boolean font, boolean typeface, boolean hook,
                               SortOrder sort, boolean reverse) {
        allAppsSelected = all;
        userAppsSelected = user;
        systemAppsSelected = system;
        injectedOnly = injected;
        disabledOnly = disabled;
        widthConfiguredOnly = viewport;
        fontConfiguredOnly = font;
        typefaceConfiguredOnly = typeface;
        hookConfiguredOnly = hook;
        sortOrder = sort != null ? sort : SortOrder.NAME;
        reverseOrder = reverse;
    }
    /** Legacy constructor: true includes system apps while user apps remain included. */
    public AppListFilterState(boolean showSystem, boolean injected, boolean viewport, boolean font) {
        this(false, true, showSystem, injected, false, viewport, font, false, false, SortOrder.NAME, false);
    }
    public static AppListFilterState defaultState() {
        return new AppListFilterState(true, false, false, false, false, false, false, false, false,
                SortOrder.NAME, false);
    }
    public static AppListFilterState noAdditionalConstraints() { return defaultState(); }
    public boolean showSystemApps() { return systemAppsSelected; }
    public boolean allAppsSelected() { return allAppsSelected; }
    public boolean userAppsSelected() { return userAppsSelected; }
    public boolean systemAppsSelected() { return systemAppsSelected; }
    public boolean injectedOnly() { return injectedOnly; }
    public boolean disabledOnly() { return disabledOnly; }
    public boolean widthConfiguredOnly() { return widthConfiguredOnly; }
    public boolean fontConfiguredOnly() { return fontConfiguredOnly; }
    public boolean typefaceConfiguredOnly() { return typefaceConfiguredOnly; }
    public boolean hookConfiguredOnly() { return hookConfiguredOnly; }
    /** True only when no configuration constraint is active. */
    public boolean allConfigurationSelected() {
        return !injectedOnly && !disabledOnly && !widthConfiguredOnly && !fontConfiguredOnly
                && !typefaceConfiguredOnly && !hookConfiguredOnly;
    }
    public boolean isDefaultSelection() {
        return allAppsSelected && allConfigurationSelected()
                && sortOrder == SortOrder.NAME && !reverseOrder;
    }
    public AppType appType() { return allAppsSelected ? AppType.ALL : systemAppsSelected && !userAppsSelected ? AppType.SYSTEM : AppType.USER; }
    public ConfigurationFilter configurationFilter() {
        int count = (injectedOnly ? 1 : 0) + (disabledOnly ? 1 : 0) + (widthConfiguredOnly ? 1 : 0)
                + (fontConfiguredOnly ? 1 : 0) + (typefaceConfiguredOnly ? 1 : 0)
                + (hookConfiguredOnly ? 1 : 0);
        if (count != 1) return ConfigurationFilter.ALL;
        if (injectedOnly) return ConfigurationFilter.INJECTED;
        if (disabledOnly) return ConfigurationFilter.DISABLED;
        if (widthConfiguredOnly) return ConfigurationFilter.VIEWPORT;
        if (fontConfiguredOnly) return ConfigurationFilter.FONT;
        return typefaceConfiguredOnly ? ConfigurationFilter.TYPEFACE : ConfigurationFilter.HOOK;
    }
    public SortOrder sortOrder() { return sortOrder; }
    public boolean reverseOrder() { return reverseOrder; }
    public AppListFilterState withAppType(AppType value) { return value == AppType.ALL ? withAllApps() : withAppTypes(value == AppType.USER, value == AppType.SYSTEM); }
    public AppListFilterState withAppTypes(boolean user, boolean system) { return new AppListFilterState(false, user, system, injectedOnly, disabledOnly, widthConfiguredOnly, fontConfiguredOnly, typefaceConfiguredOnly, hookConfiguredOnly, sortOrder, reverseOrder); }
    public AppListFilterState withAllApps() { return new AppListFilterState(true, false, false, injectedOnly, disabledOnly, widthConfiguredOnly, fontConfiguredOnly, typefaceConfiguredOnly, hookConfiguredOnly, sortOrder, reverseOrder); }
    public AppListFilterState withConfigurationFilter(ConfigurationFilter value) { return new AppListFilterState(allAppsSelected, userAppsSelected, systemAppsSelected, value == ConfigurationFilter.INJECTED, value == ConfigurationFilter.DISABLED, value == ConfigurationFilter.VIEWPORT, value == ConfigurationFilter.FONT, value == ConfigurationFilter.TYPEFACE, value == ConfigurationFilter.HOOK, sortOrder, reverseOrder); }
    public AppListFilterState withConfiguration(boolean injected, boolean viewport, boolean font) { return withConfiguration(injected, viewport, font, typefaceConfiguredOnly, hookConfiguredOnly); }
    public AppListFilterState withConfiguration(boolean injected, boolean viewport, boolean font, boolean typeface, boolean hook) { return withConfiguration(injected, disabledOnly, viewport, font, typeface, hook); }
    public AppListFilterState withConfiguration(boolean injected, boolean disabled, boolean viewport, boolean font, boolean typeface, boolean hook) { return new AppListFilterState(allAppsSelected, userAppsSelected, systemAppsSelected, injected, disabled, viewport, font, typeface, hook, sortOrder, reverseOrder); }
    public AppListFilterState withSortOrder(SortOrder value) { return new AppListFilterState(allAppsSelected, userAppsSelected, systemAppsSelected, injectedOnly, disabledOnly, widthConfiguredOnly, fontConfiguredOnly, typefaceConfiguredOnly, hookConfiguredOnly, value, reverseOrder); }
    public AppListFilterState withReverseOrder(boolean value) { return new AppListFilterState(allAppsSelected, userAppsSelected, systemAppsSelected, injectedOnly, disabledOnly, widthConfiguredOnly, fontConfiguredOnly, typefaceConfiguredOnly, hookConfiguredOnly, sortOrder, value); }
}

File: app/src/main/java/com/dpis/module/applist/AppListFilterState.kt
package com.dpis.module.applist

/** Immutable app-list filter selection. Empty sets mean that the corresponding group is unrestricted. */
class AppListFilterState(
    selectedAppTypes: Set<AppType>,
    selectedConfigurationFilters: Set<ConfigurationFilter>,
    val sortOrder: SortOrder,
    private val reverse: Boolean,
) {
    enum class AppType { USER, SYSTEM, ALL }

    enum class ConfigurationFilter { ALL, INJECTED, DISABLED, VIEWPORT, FONT, TYPEFACE, HOOK }

    enum class SortOrder { NAME, UPDATED, INSTALLED }

    val selectedAppTypes: Set<AppType> = selectedAppTypes
        .filter { it != AppType.ALL }
        .toSet()
    val selectedConfigurationFilters: Set<ConfigurationFilter> = selectedConfigurationFilters
        .filter { it != ConfigurationFilter.ALL }
        .toSet()

    fun allAppsSelected(): Boolean = selectedAppTypes.isEmpty()

    fun userAppsSelected(): Boolean = AppType.USER in selectedAppTypes

    fun systemAppsSelected(): Boolean = AppType.SYSTEM in selectedAppTypes

    fun showSystemApps(): Boolean = systemAppsSelected()

    fun injectedOnly(): Boolean = ConfigurationFilter.INJECTED in selectedConfigurationFilters

    fun disabledOnly(): Boolean = ConfigurationFilter.DISABLED in selectedConfigurationFilters

    fun widthConfiguredOnly(): Boolean =
        ConfigurationFilter.VIEWPORT in selectedConfigurationFilters

    fun fontConfiguredOnly(): Boolean = ConfigurationFilter.FONT in selectedConfigurationFilters

    fun typefaceConfiguredOnly(): Boolean =
        ConfigurationFilter.TYPEFACE in selectedConfigurationFilters

    fun hookConfiguredOnly(): Boolean = ConfigurationFilter.HOOK in selectedConfigurationFilters

    fun allConfigurationSelected(): Boolean = selectedConfigurationFilters.isEmpty()

    fun isDefaultSelection(): Boolean =
        allAppsSelected() && allConfigurationSelected() && sortOrder == SortOrder.NAME && !reverse

    /** Legacy single-value view used by saved-instance-state compatibility. */
    fun appType(): AppType = when {
        allAppsSelected() -> AppType.ALL
        systemAppsSelected() && !userAppsSelected() -> AppType.SYSTEM
        else -> AppType.USER
    }

    fun sortOrder(): SortOrder = sortOrder

    fun reverseOrder(): Boolean = reverse

    fun toggleAppType(type: AppType): AppListFilterState = when (type) {
        AppType.ALL -> selectAllAppTypes()
        AppType.USER, AppType.SYSTEM -> {
            val next = if (type in selectedAppTypes) {
                selectedAppTypes - type
            } else {
                selectedAppTypes + type
            }
            copy(selectedAppTypes = next)
        }
    }

    fun selectAllAppTypes(): AppListFilterState = copy(selectedAppTypes = emptySet())

    fun selectAppTypes(types: Set<AppType>): AppListFilterState = copy(selectedAppTypes = types)

    fun toggleConfiguration(filter: ConfigurationFilter): AppListFilterState {
        if (filter == ConfigurationFilter.ALL) {
            return clearConfigurationFilters()
        }
        val next = if (filter in selectedConfigurationFilters) {
            selectedConfigurationFilters - filter
        } else {
            selectedConfigurationFilters + filter
        }
        return copy(selectedConfigurationFilters = next)
    }

    fun clearConfigurationFilters(): AppListFilterState =
        copy(selectedConfigurationFilters = emptySet())

    fun withSortOrder(value: SortOrder): AppListFilterState = copy(sortOrder = value)

    fun withReverseOrder(value: Boolean): AppListFilterState = copy(reverse = value)

    private fun copy(
        selectedAppTypes: Set<AppType> = this.selectedAppTypes,
        selectedConfigurationFilters: Set<ConfigurationFilter> = this.selectedConfigurationFilters,
        sortOrder: SortOrder = this.sortOrder,
        reverse: Boolean = this.reverse,
    ): AppListFilterState = AppListFilterState(
        selectedAppTypes,
        selectedConfigurationFilters,
        sortOrder,
        reverse,
    )

    companion object {
        @JvmStatic
        fun defaultState(): AppListFilterState = AppListFilterState(
            emptySet(),
            emptySet(),
            SortOrder.NAME,
            false,
        )

        @JvmStatic
        fun noAdditionalConstraints(): AppListFilterState = defaultState()
    }
}
